integration-libs/opf/checkout/components/opf-checkout-billing-address-form/opf-checkout-billing-address-form.service.ts

```diff
@@ -0,0 +1,183 @@
+import { ActiveCartFacade, Cart } from '@spartacus/cart/base/root';
+import {
+  CheckoutBillingAddressFacade,
+  CheckoutDeliveryAddressFacade,
+  CheckoutPaymentFacade,
+} from '@spartacus/checkout/base/root';
+import {
+  Address,
+  Country,
+  GlobalMessageService,
+  GlobalMessageType,
+  HttpErrorModel,
+  UserPaymentService,
+} from '@spartacus/core';
+import {
+  BehaviorSubject,
+  combineLatest,
+  EMPTY,
+  Observable,
+  throwError,
+} from 'rxjs';
+import {
+  catchError,
+  filter,
+  finalize,
+  map,
+  shareReplay,
+  switchMap,
+  take,
+  tap,
+} from 'rxjs/operators';
+import { OpfCheckoutPaymentWrapperService } from '../opf-checkout-payment-wrapper';
+
+@Injectable()
+export class OpfCheckoutBillingAddressFormService {
+  protected readonly billingAddressSub = new BehaviorSubject<
+    Address | undefined
+  >(undefined);
+  protected readonly isLoadingAddressSub = new BehaviorSubject(false);
+  protected readonly isSameAsDeliverySub = new BehaviorSubject(true);
+  protected billingAddressId: string | undefined;
+
+  billingAddress$ = this.billingAddressSub.asObservable();
+  isLoadingAddress$ = this.isLoadingAddressSub.asObservable();
+  isSameAsDelivery$ = this.isSameAsDeliverySub.asObservable();
+
+  constructor(
+    protected checkoutDeliveryAddressFacade: CheckoutDeliveryAddressFacade,
+    protected checkoutBillingAddressFacade: CheckoutBillingAddressFacade,
+    protected userPaymentService: UserPaymentService,
+    protected checkoutPaymentService: CheckoutPaymentFacade,
+    protected activeCartService: ActiveCartFacade,
+    protected globalMessageService: GlobalMessageService,
+    protected opfService: OpfCheckoutPaymentWrapperService
+  ) {}
+
+  getCountries(): Observable<Country[]> {
+    return this.userPaymentService.getAllBillingCountries().pipe(
+      tap((countries) => {
+        if (Object.keys(countries).length === 0) {
+          this.userPaymentService.loadBillingCountries();
+        }
+      }),
+      // we want to share data with the address form and prevent loading data twice
+      shareReplay(1)
+    );
+  }
+
+  getAddresses(): void {
+    this.isLoadingAddressSub.next(true);
+
+    combineLatest([this.getDeliveryAddress(), this.getPaymentAddress()])
+      .pipe(take(1))
+      .subscribe(
+        ([deliveryAddress, paymentAddress]: [
+          Address | undefined,
+          Address | undefined
+        ]) => {
+          if (!paymentAddress && !!deliveryAddress) {
+            this.setBillingAddress(deliveryAddress);
+            this.billingAddressSub.next(deliveryAddress);
+          }
+
+          if (!!paymentAddress && !!deliveryAddress) {
+            this.billingAddressId = paymentAddress.id;
+            this.billingAddressSub.next(paymentAddress);
+            this.isSameAsDeliverySub.next(false);
+          }
+
+          this.isLoadingAddressSub.next(false);
+        }
+      );
+  }
+
+  putDeliveryAddressAsPaymentAddress(): void {
```

Comment:
[MEDIUM]

`put...` vs `set...` ? asking for consistency.

Reply:
[OUTDATED]